### PR TITLE
Attendance feature

### PR DESCRIPTION
**Pull request by tohlh**

### CODE REVIEW

File: src/main/java/seedu/address/model/group/Group.java [hunk 1]
```diff
@@ -11,15 +14,19 @@ public class Group {
 
     public static final String MESSAGE_CONSTRAINTS = "Groups names should be in correct format with 2 digit number. "
         + "E.g. g/TUT04, g/LAB10, g/REC09. Link should be a valid Telegram invite link.";
-
+    public static final String WEEK_MESSAGE_CONSTRAINTS = "Week number should be 2 digit integer. ";
+    public static final String ATTENDANCE_MESSAGE_CONSTRAINTS = "Attendance should be A or P. ";
```
Comment (jocelyn-soh) [inline src/main/java/seedu/address/model/group/Group.java:18]: I like that the error messages are clear!

File: src/main/java/seedu/address/model/group/Group.java [hunk 2]
```diff
@@ -31,18 +38,49 @@ public Group(String groupName) {
         checkArgument(isValidGroupName(groupName), MESSAGE_CONSTRAINTS);
         this.groupName = groupName;
         this.telegramLink = "";
+        this.attendance = new ArrayList<>();
+        for (int i = 0; i < 13; i++) {
```
Comment (jocelyn-soh) [inline src/main/java/seedu/address/model/group/Group.java:42]: I think it's smart to restrict attendance to 13 weeks.

**Review by jocelyn-soh — COMMENTED** [review]
LGTM

**Comment by doyelee0313** [comment]
LGTM

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java [hunk 3]
```diff
@@ -43,7 +43,8 @@ class JsonAdaptedPerson {
     public JsonAdaptedPerson(@JsonProperty("name") String name, @JsonProperty("phone") String phone,
             @JsonProperty("email") String email, @JsonProperty("year") String year,
             @JsonProperty("telegram") String telegram, @JsonProperty("major") String major,
-            @JsonProperty("remark") String remark, @JsonProperty("groups") List<String> groups) {
+            @JsonProperty("remark") String remark,
+            @JsonProperty("groups") List<JsonAdaptedPersonGroupAttendance> groups) {
```
Comment (doyelee0313) [inline src/main/java/seedu/address/storage/JsonAdaptedPerson.java:47]: Good to have another Json fo Attendance

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java [hunk 4]
```diff
@@ -68,8 +69,7 @@ public JsonAdaptedPerson(Person source) {
         telegram = source.getTelegram().value;
         remark = source.getRemark().value;
         groups.addAll(source.getGroups().stream()
-                .map(JsonAdaptedGroup::new)
-                .map(JsonAdaptedGroup::getGroupName)
+                .map(JsonAdaptedPersonGroupAttendance::new)
```
Comment (doyelee0313) [inline src/main/java/seedu/address/storage/JsonAdaptedPerson.java:72]: Good to show the attendance in json file

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java [hunk 5]
```diff
@@ -35,9 +35,8 @@ public class JsonAdaptedPersonTest {
     private static final String VALID_MAJOR = BENSON.getMajor().toString();
     private static final String VALID_TELEGRAM = BENSON.getTelegram().toString();
     private static final String VALID_REMARK = BENSON.getRemark().toString();
-    private static final List<String> VALID_GROUPS = BENSON.getGroups().stream()
-            .map(JsonAdaptedGroup::new)
-            .map(JsonAdaptedGroup::getGroupName)
+    private static final List<JsonAdaptedPersonGroupAttendance> VALID_GROUPS = BENSON.getGroups().stream()
+            .map(JsonAdaptedPersonGroupAttendance::new)
```
Comment (doyelee0313) [inline src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java:39]: good to omit the group name and make attendance json

File: src/main/java/seedu/address/logic/commands/MarkAttendanceCommand.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Marks attendance for a specified person in a group.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + PREFIX_GROUP + "GROUP_NAME "
+            + PREFIX_WEEK + "WEEK_NUMBER (week 1 to week 13)"
+            + PREFIX_ATTENDANCE + "ABSENT_OR_PRESENT (A for absent P for present)"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_GROUP + "TUT04 "
+            + PREFIX_WEEK + "3 "
+            + PREFIX_ATTENDANCE + "A ";
```
Comment (doyelee0313) [inline src/main/java/seedu/address/logic/commands/MarkAttendanceCommand.java:32]: good to show the example command words

File: src/main/java/seedu/address/logic/commands/MarkAttendanceCommand.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    public static final String MESSAGE_SUCCESS = "Attendance marked";
+    public static final String MESSAGE_WEEK_NUMBER_INVALID = "Week number is wrong (between 1 and 13)";
+    public static final String MESSAGE_ATTENDANCE_INVALID = "Attendance format is wrong (A for absent P for present)";
+    public static final String MESSAGE_GROUP_NOT_FOUND = "Group is not found in the person";
+
+    private final Index index;
+    private final Group group;
+    private final int week;
+    private final String attendance;
```
Comment (doyelee0313) [inline src/main/java/seedu/address/logic/commands/MarkAttendanceCommand.java:42]: good to have the attendance in a string list

**Review by doyelee0313 — COMMENTED** [review]
LGTM

**Pull request merged**